shell/run_shell_cmd.go

```diff
@@ -86,12 +105,17 @@ func RunShellCommandWithOutput(terragruntOptions *options.TerragruntOptions, wor
 	return &cmdOutput, errors.WithStackTrace(err)
 }
 
-func toEnvVarsList(envVarsAsMap map[string]string) []string {
+func toEnvVarsList(envVarsAsMap map[string]string, terragruntOptions *options.TerragruntOptions) ([]string, error) {
 	envVarsAsList := []string{}
+	include := terragruntOptions.GetIncludeConfig()
 	for key, value := range envVarsAsMap {
-		envVarsAsList = append(envVarsAsList, fmt.Sprintf("%s=%s", key, value))
+		pvalue, err := ResolveTerragruntConfigString(value, include, terragruntOptions)
+		if err != nil {
+			return nil, err
+		}
+		envVarsAsList = append(envVarsAsList, fmt.Sprintf("%s=%s", key, pvalue))
 	}
-	return envVarsAsList
+	return envVarsAsList, nil
```

Comment:
Again, same thing here. This method should not know anything about interpolations. All the interpolation process should happen in a single location before the `shell` package is ever called.

Reply:
I doubt that I understood the code of terragrunt that thoroughly so that I could be the one to add new abstraction layers here. Perhaps you can come up with a proposal here.

Reply:
Well, to make a major changes to how interpolations work and when they are processed, I think understanding the flow fully is a requirement.

The way I'd approach it is to look at the places these `run_shell_cmd.go` functions are used. Is there a common path they are going through? What's the order of events? Running init, running before hooks, executing Terraform, running after hooks, etc. When are interpolations  processed now (i.e., before any of those events)? And when do you want them to be processed (i.e., at each event?).